shepherd/shepherd.py

```diff
@@ -0,0 +1,438 @@
+from multiprocessing import Process
+import json
+import multiprocessing
+import os
+import signal
+import subprocess
+import sys
+import threading
+import time
+import yaml
+
+# --- config_loader.py ---
+
+
+def load_and_preprocess_config(filepath):
+    """Loads and preprocesses configuration from a YAML file."""
+    with open(filepath, 'r') as file:
+        config = yaml.safe_load(file)
+
+    preprocess_config(config, filepath)
+
+    print(f"DEBUG: Loaded and preprocessed config from {filepath}")
```

Comment:
Since this is an independent module, you may want to use the standard python logging module.